### PR TITLE
Validated form fields

### PR DESCRIPTION
**Pull request by MaddieLvrg**
The submit buttons are disabled until all forms are completed. Will show
an error message if something is incorrect.

### CODE REVIEW

**Comment by kjrcda** [comment]
The spans is the wrong way to do this, on the main page there is already an error handler, just need to hook into that. Same idea for the register page. This also leads to not disabling the buttons, when a user clocks the button and has invalid input then they should be notified.

File: webapp/public/javascripts/controllers.js [hunk 1]
```diff
@@ -26,7 +26,13 @@ lunchrControllers.controller('MainPageController', ['$scope', '$http', '$state',
                 return;
             }
 
-            $http.post('/api/users/authenticate', {email: $scope.email, password: $scope.password})
+            if(/^[_a-z0-9]+(\.[_a-z0-9]+)*@[a-z0-9-]+(\.[a-z0-9-]+)*(\.[a-z]{2,4})$/.test($scope.email) == false)
+            {
+                $scope.errorMessages = "Incorrect Email";
```
Comment (paymog) [inline webapp/public/javascripts/controllers.js]: Incorrect or invalid?

File: webapp/routes/api.js [hunk 2]
```diff
@@ -55,17 +55,34 @@ router.post('/users/register', function (req, res, next) {
 });
 
 router.post('/users/authenticate', function (req, res, next) {
-    User.findOne({email: req.body.email, password: req.body.password}, function (error, user) {
+    User.find({email: req.body.email}, function (error, users) {
```
Comment (paymog) [inline webapp/routes/api.js]: Why was this changed? There should only ever be a single person with an email, right?

File: webapp/routes/api.js [hunk 3]
```diff
@@ -55,17 +55,34 @@ router.post('/users/register', function (req, res, next) {
...
+                console.log("finds match");
+                users[i].password = null;
+                res.json(users[i]);
+                //technically don't need these 2 lines, but here in case
+                found = true;
+                break;
+            }
+        }
+
+        if(found == false) {
```
Comment (paymog) [inline webapp/routes/api.js]: use `if(!found)`

File: webapp/routes/api.js (hunk 2 above)
Comment (kjrcda) [inline webapp/routes/api.js]: This was changed so that we can tell if there is a valid email but incorrect password entered

Comment (paymog) [inline webapp/routes/api.js]: but, still do findOne and just compare passwords?

Comment (paymog) [inline webapp/routes/api.js]: like, semantically, `find` should only ever return an array of length 1 or 0

File: webapp/public/javascripts/controllers.js [hunk 4]
```diff
@@ -66,8 +72,14 @@ lunchrControllers.controller('RegisterController', ['$scope', '$http', '$state',
...
 
+            if(/^[_a-z0-9]+(\.[_a-z0-9]+)*@[a-z0-9-]+(\.[a-z0-9-]+)*(\.[a-z]{2,4})$/.test($scope.email) == false)
+            {
+                $scope.errorMessages = "Incorrect Email";
+                return;
+            }
+
             $http.post('/api/users/register', {
-                email: $scope.email,
+                email: $scope.email.toLowerCase(),
```
Comment (paymog) [inline webapp/public/javascripts/controllers.js]: We should lowercase on the serverside. No here. That way the ios client works with this concept too.

File: webapp/routes/api.js [hunk 5]
```diff
@@ -63,9 +63,14 @@ router.post('/users/authenticate', function (req, res, next) {
         if (!user) {
             return next(new Error("User not found"));
         }
+
+        if(user.password != req.body.password) {
+            return next(new Error("Incorrect Password"));
+        }
+
         user.password = null;
         res.json(user);
```
Comment (paymog) [inline webapp/routes/api.js:73]: do we have to undo this null set here?

File: webapp/routes/api.js [hunk 6]
```diff
@@ -36,7 +36,7 @@ router.post('/users/register', function (req, res, next) {
 
         // create and save the user
         var user = new User({
-            email: req.body.email,
+            email: req.body.email.toLowerCase(),
```
Comment (paymog) [inline webapp/routes/api.js:39]: we're not lowercasing in the find statement above

File: webapp/routes/api.js (hunk 5 above)
Comment (kjrcda) [inline webapp/routes/api.js:73]: I don't think so. The behavior is ok in the api but wasn't in socket. I will add the resetting of the password just in case though.
